sympy/combinatorics/homomorphisms.py

```diff
@@ -417,3 +419,96 @@ def block_homomorphism(group, blocks):
     images = {g: Permutation([b[p[i]^g] for i in identity]) for g in group.generators}
     H = GroupHomomorphism(group, codomain, images)
     return H
+
+def group_isomorphism(G, H):
+    '''
+    Compute an isomorphism between 2 given groups.
+
+    Arguments:
+        G (a finite `FpGroup` or a `PermutationGroup`) -- First group
+        H (a finite `FpGroup` or a `PermutationGroup`) -- Second group
+
+    Returns:
+    If the groups are not isomorphic, `None` is returned.
+    If the groups are isomorphic, a corresponding `GroupHomomorphism` object is returned.
+
+    Summary:
+    Uses the approach suggested by Robert Tarjan to compute the isomorphism between two groups.
+    First, the set of generators are mapped with the elements of `H` and we check if the mapping induces an isomorphism.
+
+    '''
+
+    if isinstance(G, FpGroup):
+        if G.order() == S.Infinity:
+            raise NotImplementedError("Isomorphism methods are not implemented for infinite groups.")
+        G = G._to_perm_group()[0]
+    if isinstance(H, FpGroup):
+        if H.order() == S.Infinity:
+            raise NotImplementedError("Isomorphism methods are not implemented for infinite groups.")
+        H = H._to_perm_group()[0]
+
+    if not isinstance(G, PermutationGroup):
+        raise TypeError("The group must be a PermutationGroup or a finite FpGroup")
+    if not isinstance(H, PermutationGroup):
+        raise TypeError("The group must be a PermutationGroup or a finite FpGroup")
+
+    if (G.order() != H.order()) and (G.is_abelian != H.is_abelian):
+        return None
+
+    # Match the generators of `G` with the subsets of H
+    gens = G.generators
+    for subset in itertools.permutations(H, len(gens)):
+        # Create a map
+        func_map = dict(zip(gens, subset))
+        counterexample = False
+        # Loop through the mapped elements and try to extend the mapping
+        # or to find a counter example.
+        while not counterexample:
+            extended_map = {}
+            for g, h in itertools.product(func_map, func_map):
+                if g*h in func_map:
+                    if func_map[g]*func_map[h] != func_map[g*h]:
+                        counterexample = True
+                        break
+                else:
+                    extended_map[g*h] = func_map[g]*func_map[h]
+
+            # Break when all the elements are mapped.
+            if len(func_map) == H.order() :
+                break
+
+            # Remove the duplicate elements in extended_map
+            # and merge the extended_map with the map.
+            image_len = len(set(func_map.values()) | set(extended_map.values()))
+            if image_len != len(func_map) + len(extended_map):
+                counterexample = True
+                break
+            func_map.update(extended_map)
+
+        if not counterexample:
+            # Returns a `GroupHomomorphism` object.
+            return homomorphism(G, H, G.generators, subset, check=False)
```

Comment:
Just realised: if `G` or `H` are `FpGroup`s, then you are returning an isomorphism between permutation groups that are isomorphic to them, not between the groups that were supplied as arguments. That is not a behavior we want. For example, in on of your tests `G = FpGroup(F, [a**3, b**3, (a*b)**2])` but `T` acts on permutations, not on elements of `G`, e.g. `a*b`.

Reply:
Oh yes. Maybe, we can just prevent the overwriting of `G` and `H`.For example, if `G` is an `FpGroup`, we define `G* = G._to_perm_grp()` and all the computations are done with respect to `G*` instead of `G`and finally if we don't find a counterexample, we return the corresponding homomorphism between `G` and `H`.

Reply:
But, if `H` is an `FpGroup` won't that be a problem because `subset` will be a subset of the `PermutationGroup` isomorphic to `H`?

Reply:
Not if you get an isomorphism between `H` and  `H._to_perm_grp` - then you could rewrite `subset` with the corresponding inverse images. You should do the same for `G` because the generators of `G._to_perm_grp` might not necessarily correspond to those of `G` in the obvious way. Have a look in `fp_groups.py`, near where `_to_perm_grp` is defined - there is a way to return a group and the corresponding isomorphism (and several functions in that section demonstrate how to go between elements of `G` and its permutation group representation).

Reply:
To add to the above (which is still to be implemented), if `compute_isomorphism` is called from `is_isomorphic` with `isomorphism = False`, then we don't need to actually call `homomorphism`. I'm not sure how much time it actually takes but if it's significant and the user doesn't need the actual isomorphism, then we shouldn't do this.

Reply:
> You should do the same for G because the generators of G._to_perm_grp might not necessarily correspond to those of G in the obvious way. 

Can't we simply use `G.generators` there? While mapping the generators, we use the generators of `_G`(which is the permutation group isomorphic to `G`), but, when we call the `homomorphism` function, we use `G.generators` there.

Reply:
> To add to the above (which is still to be implemented), if compute_isomorphism is called from is_isomorphic with isomorphism = False, then we don't need to actually call homomorphism

Then we don't actually need the isomorphism method, right? We can have another argument `compute` in the `group_isomorphism` method and when `compute` is set to `False` a boolean is returned.

Reply:
> Can't we simply use `G.generators` there? 

You can but currently that's not what you are doing because you are rewriting `G` with its permutation group. And `H` will actually need to use the isomorphism to its permutation group if it's an `FpGroup`.

Reply:
> Then we don't actually need the is_isomorphic method, right? We can have another argument compute in the `group_isomorphism` method and when compute is set to False a boolean is returned.

I was thinking this. I'm not sure whether it would be better to call it `is_isomorphic` or `group_isomorphism` in that case. Both can make sense. I suppose the latter may work somewhat better.